### PR TITLE
[4.0] Add Count items to Tags Overview

### PR DESCRIPTION
**Pull request by pe7er**
Pull Request to **improve** the **Tags** component by adding a **new "Tagged Items" column** that shows a **counter** to offer an insight in how often certain tags are used. The column is sortable so you can easily get an overview about the most or less used tags.
It's PR https://github.com/joomla/joomla-cms/pull/23562 re-done for Joomla 4. 

### Summary of Changes
In the Back-end under Components > Tags you'll see an overview of all created Tags.
However, you don't have any information which tags are actually used, and how often.

tagged-items-before

### Testing Instructions
-  Add some tags to some items (Articles or other items).
-  Check the Components > Tags overview
-  Install the patch
-  Check again the Components > Tags overview

### Expected result
A new column "**Tagged Items**" is available. You can **click** on the "Tagged Items" header **to order** the column. You can select the order also via the "Sort Table By:" dropdown on the top right.

tagged-items-after

### CODE REVIEW

**Comment by brianteeman** [comment]
I have tested this item :white_check_mark: successfully on 12f4d7b2a8e897fe2424f2ffaff1c95ac560ea7e<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/24663">issues.joomla.org/tracker/joomla-cms/24663</a>.</sub>

**Comment by ghazal** [comment]
I have tested this item :white_check_mark: successfully on 12f4d7b2a8e897fe2424f2ffaff1c95ac560ea7e<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/24663">issues.joomla.org/tracker/joomla-cms/24663</a>.</sub>

**Comment by ghost** [comment]
Status "Ready To Commit".

File: administrator/components/com_tags/Model/TagsModel.php [hunk 1]
```diff
@@ -144,6 +145,14 @@ protected function getListQuery()
 		$query->from('#__tags AS a')
 			->where('a.alias <> ' . $db->quote('root'));
 
+		// Count Items
+		$subQueryCountTaggedItems = $db->getQuery(true);
+		$subQueryCountTaggedItems
+			->select('COUNT(' . $db->quoteName('tag_map.content_item_id') . ')')
+			->from($db->quoteName('#__contentitem_tag_map', 'tag_map'))
+			->where($db->quoteName('tag_map.tag_id') . '= ' . $db->quoteName('a.id'));
+		$query->select('(' . $subQueryCountTaggedItems->__toString() . ') AS countTaggedItems');
```
Comment (wilsonge) [inline administrator/components/com_tags/Model/TagsModel.php]: You shouldn't ever call the magic method `__toString()`. Just typehint it with `(string) $subQueryCountTaggedItems`

Comment (pe7er) [reply administrator/components/com_tags/Model/TagsModel.php]: @wilsonge Ok, thanks for the warning. I'll correct it!

**Comment by pe7er** [comment]
@wilsonge Could you please check if I fixed " magic method __toString() to typehint (string)" correctly in 
administrator/components/com_tags/Model/TagsModel.php ?

**Comment by pe7er** [comment]
@brianteeman + @ghazal Thanks for testing!

File: administrator/components/com_tags/tmpl/tags/default.php [hunk 2]
```diff
@@ -115,6 +115,8 @@
 								<?php echo HTMLHelper::_('searchtools.sort', 'JGRID_HEADING_LANGUAGE', 'a.language', $this->state->get('list.direction'), $this->state->get('list.ordering')); ?>
 							</th>
 						<?php endif; ?>
+                        <th scope="col" style="width:10%" class="d-none d-md-table-cell text-center">
```
Comment (zero-24) [inline administrator/components/com_tags/tmpl/tags/default.php]: spaces to tabs please

File: administrator/components/com_tags/tmpl/tags/default.php [hunk 3]
```diff
@@ -236,6 +238,11 @@
 								<?php echo LayoutHelper::render('joomla.content.language', $item); ?>
 							</td>
 						<?php endif; ?>
+                        <td class="small d-none d-md-table-cell text-center">
```
Comment (zero-24) [inline administrator/components/com_tags/tmpl/tags/default.php]: spaces to tabs please

**Review by zero-24 — CHANGES_REQUESTED** [review]

File: administrator/components/com_tags/tmpl/tags/default.php (hunk 2 above)
Comment (pe7er) [reply administrator/components/com_tags/tmpl/tags/default.php]: @zero-24 Sorry... I fixed it!

File: administrator/components/com_tags/tmpl/tags/default.php (hunk 3 above)
Comment (pe7er) [reply administrator/components/com_tags/tmpl/tags/default.php]: Fixed!

**Review by zero-24 — APPROVED** [review]

**Comment by ghazal** [comment]
I have tested this item :white_check_mark: successfully on 9f32ae16191069a224b212da04612490c6d9f9c6<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/24663">issues.joomla.org/tracker/joomla-cms/24663</a>.</sub>

**Comment by wilsonge** [comment]
Looks better!

File: administrator/components/com_tags/Model/TagsModel.php [hunk 4]
```diff
@@ -144,6 +145,14 @@ protected function getListQuery()
 		$query->from('#__tags AS a')
 			->where('a.alias <> ' . $db->quote('root'));
 
+		// Count Items
+		$subQueryCountTaggedItems = $db->getQuery(true);
+		$subQueryCountTaggedItems
+			->select('COUNT(' . $db->quoteName('tag_map.content_item_id') . ')')
+			->from($db->quoteName('#__contentitem_tag_map', 'tag_map'))
+			->where($db->quoteName('tag_map.tag_id') . '= ' . $db->quoteName('a.id'));
```
Comment (roland-d) [inline administrator/components/com_tags/Model/TagsModel.php]: Space before `=`

File: administrator/components/com_tags/Model/TagsModel.php [hunk 5]
```diff
@@ -144,6 +145,14 @@ protected function getListQuery()
...
+		$query->select('(' . (string) $subQueryCountTaggedItems . ') AS countTaggedItems');
```
Comment (roland-d) [inline administrator/components/com_tags/Model/TagsModel.php]: Quote name `countTaggedItems`

File: administrator/components/com_tags/Model/TagsModel.php (hunk 4 above)
Comment (Quy) [inline administrator/components/com_tags/Model/TagsModel.php]:
```suggestion
			->where($db->quoteName('tag_map.tag_id') . ' = ' . $db->quoteName('a.id'));
```

**Comment by roland-d** [comment]
Removing RTC until issues are resolved

**Comment by pranotiTechjoomla** [comment]
I have tested this item :white_check_mark: successfully on 9c9a8de54744a482d5e61ef782635495ee8599d8<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/24663">issues.joomla.org/tracker/joomla-cms/24663</a>.</sub>

**Comment by ttplpoojak** [comment]
I have tested this item :white_check_mark: successfully on d795774844def2d2e6b500b9a866138e6f0f0a0b<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/24663">issues.joomla.org/tracker/joomla-cms/24663</a>.</sub>

**Comment by joeykrop** [comment]
I have tested this item :white_check_mark: successfully on d795774844def2d2e6b500b9a866138e6f0f0a0b<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/24663">issues.joomla.org/tracker/joomla-cms/24663</a>.</sub>

**Comment by alikon** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/24663">issues.joomla.org/tracker/joomla-cms/24663</a>.</sub>

**Comment by Quy** [comment]
There are coding style changes.

**Comment by zero-24** [comment]
Back to Pending until @pe7er found to time to fix the issues mention above.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/24663">issues.joomla.org/tracker/joomla-cms/24663</a>.</sub>

**Comment by pe7er** [comment]
Thank you all for testing & your suggestions for improvements!
 
As I did not have the branch for this PR in my local git repo (hence it says "from unknown repository"),
I redid the PR with the suggested improvements.

New PR: https://github.com/joomla/joomla-cms/pull/28831

I'm closing this one...

**Pull request closed without merging**